File: taskcenter/src/main/java/com/iwdael/taskcenter/core/NodeUnitary.java
package com.iwdael.taskcenter.core;

import com.iwdael.taskcenter.interfaces.CloseCreator;
import com.iwdael.taskcenter.interfaces.DisperseCreator;
import com.iwdael.taskcenter.interfaces.Interceptor;
import com.iwdael.taskcenter.interfaces.Mapper;
import com.iwdael.taskcenter.interfaces.UnitCreator;
import com.iwdael.taskcenter.task.TaskUnitary;

/**
 * @author : iwdael
 * @mail : iwdael@outlook.com
 * @project : https://github.com/iwdael/TaskCenter
 */
public class NodeUnitary<P, S> extends Node<P, TaskUnitary<P, S>, UnitCreator<P, S>> {
    public NodeUnitary(Class<?> type, UnitCreator<P, S> creator) {
        super(type, creator);
    }

    @Override
    protected TaskUnitary<P, S> make(P pre) {
        return creator.create(pre);
    }

    public <D> NodeDisperser<S, D> append(DisperseCreator<S, D> disperseCreator) {
        NodeDisperser<S, D> node = new NodeDisperser<>(type, disperseCreator);
        this.next = node.asNode();
        this.next.index = this.index + 1;
        this.next.pre = this.asNode();
        return node;
    }

    public <D> NodeUnitary<S, D> append(UnitCreator<S, D> unitCreator) {
        NodeUnitary<S, D> node = new NodeUnitary<>(type, unitCreator);
        this.next = node.asNode();
        this.next.index = this.index + 1;
        this.next.pre = this.asNode();
        return node;
    }


    public NodeClosure<S> append(CloseCreator<S> closeCreator) {
        NodeClosure<S> node = new NodeClosure<>(type, closeCreator);
        this.next = node.asNode();
        this.next.index = this.index + 1;
        this.next.pre = this.asNode();
        return node;
    }

    public <D> NodeMapper<S, D> append(Mapper<S, D> mapper) {
        NodeMapper<S, D> node = new NodeMapper<>(type, mapper);
        this.next = node.asNode();
        this.next.index = this.index + 1;
        this.next.pre = this.asNode();
        return node;
    }

    public <D> NodeInterceptor<S, D> append(Interceptor<S, D> interceptor) {
        NodeInterceptor<S, D> node = new NodeInterceptor<>(type, interceptor);
        this.next = node.asNode();
        this.next.index = this.index + 1;
        this.next.pre = this.asNode();
        return node;
    }
}

File: taskcenter/src/main/java/com/iwdael/taskcenter/core/NodeUnitary.kt
package com.iwdael.taskcenter.core

import com.iwdael.taskcenter.creator.CloseCreator
import com.iwdael.taskcenter.creator.DisperseCreator
import com.iwdael.taskcenter.creator.UnitCreator
import com.iwdael.taskcenter.operators.Interceptor
import com.iwdael.taskcenter.operators.Mapper

class NodeUnitary<PRE : Any, SRC : Any>(private val type: Class<*>, private val creator: UnitCreator<PRE, SRC>) : Node<PRE, Task<PRE, SRC>>() {
    override fun make(src: PRE) = creator.create(src)
    fun <DST : Any> append(disperseCreator: DisperseCreator<SRC, DST>): NodeDisperser<SRC, DST> {
        val n = NodeDisperser(type, disperseCreator)
        this.next = n.asNode()
        this.next!!.index = this.index + 1
        this.next!!.pre = this.asNode()
        return n
    }

    fun <DST : Any> append(unitCreator: UnitCreator<SRC, DST>): NodeUnitary<SRC, DST> {
        val n = NodeUnitary(type, unitCreator)
        this.next = n.asNode()
        this.next!!.index = this.index + 1
        this.next!!.pre = this.asNode()
        return n
    }

    fun <DST : Any> append(interceptor: Interceptor<SRC, DST>): NodeInterceptor<SRC, DST> {
        val n = NodeInterceptor(type, interceptor)
        this.next = n.asNode()
        this.next!!.index = this.index + 1
        this.next!!.pre = this.asNode()
        return n
    }

    fun append(closeCreator: CloseCreator<SRC>): NodeClosure<SRC> {
        val n = NodeClosure(type, closeCreator)
        this.next = n.asNode()
        this.next!!.index = this.index + 1
        this.next!!.pre = this.asNode()
        return n
    }

    fun <DST : Any> append(mapper: Mapper<SRC, DST>): NodeMapper<SRC, DST> {
        val n = NodeMapper(type, mapper)
        this.next = n.asNode()
        this.next!!.index = this.index + 1
        this.next!!.pre = this.asNode()
        return n
    }

}